tests/component.js

```diff
@@ -0,0 +1,351 @@
+		// but keep them separate in code, because they need different setup.
+		describe( 'property', () => {
+			[ {
+				property: 'value',
+				value: 'foo'
+			}, {
+				property: 'type',
+				value: 'inline'
+			}, {
+				property: 'editorUrl',
+				value: 'https://cdn.ckeditor.com/4.10.0/basic-all/ckeditor.js'
+			}, {
+				property: 'config',
+				value: {}
+			}, {
+				property: 'tagName',
+				value: 'div'
+			}, {
+				property: 'readOnly',
+				value: true
+			} ].forEach( ( { property, value } ) => {
+				setPropsForTestGroup( { [ property ]: value } );
+
+				it( `"${ property }" should be configurable`, () => {
+					expect( component[ property ] ).to.equal( value );
+				} );
+			} );
+		} );
+
+		[ {
+			readOnly: true,
+			config: { readOnly: false }
+		}, {
+			readOnly: false,
+			config: { readOnly: true }
+		} ].forEach( ( { readOnly, config } ) => {
+			describe( `when component.readOnly = ${ readOnly } and config.readOnly = ${ config.readOnly }`, () => {
+				setPropsForTestGroup( { readOnly, config } );
+
+				it( 'should use component.readOnly', () => {
+					expect( spies.replace.lastCall.args[ 1 ] ).to.include( { readOnly } );
+				} );
+			} );
+		} );
+
+		describe( 'when editor type', () => {
+			[ {
+				type: 'unset',
+				method: 'replace'
+			}, {
+				type: 'classic',
+				method: 'replace'
+			}, {
+				type: 'inline',
+				method: 'inline'
+			} ].forEach( ( { type, method } ) => {
+				describe( type === 'unset' ? 'unset' : `set to "${ type }"`, () => {
+					const config = { foo: 'bar' };
+
+					if ( type !== 'unset' ) {
+						setPropsForTestGroup( { type } );
+					} else {
+						type = 'classic';
+					}
+
+					setPropsForTestGroup( { config } );
+
+					it( `"component.type" should be "${ type }"`, () => {
+						expect( component.type ).to.equal( type );
+					} );
+
+					it( `should call "CKEDITOR.${ method }" with given config`, () => {
+						sinon.assert.calledOnce( spies[ method ] );
+
+						expect( spies[ method ].lastCall.args[ 1 ] ).to.include( config );
+					} );
+				} );
+			} );
+
+			describe( 'set to invalid value', () => {
+				it( 'should be disallowed by validator', () => {
+					expect( component.$options.props.type.validator( 'foo' ) ).to.be.false;
+				} );
+			} );
+		} );
+	} );
+
+	describe( 'events', () => {
+		[
+			'input',
+			'focus',
+			'blur'
+		].forEach( evtName => {
+			let getDataMock, spy, editorEvtName;
+
+			beforeEach( () => {
+				editorEvtName = evtName;
+				spy = sinon.spy();
+
+				if ( evtName === 'input' ) {
+					editorEvtName = 'change';
+					getDataMock = sinon.stub( component.instance, 'getData' ).returns( 'foo' );
+				}
+
+				component.$on( evtName, spy );
+				component.instance.fire( editorEvtName );
+			} );
+
+			afterEach( () => {
+				if ( getDataMock ) {
+					getDataMock.restore();
+					getDataMock = null;
+				}
+			} );
+
+			it( `should emit "${ evtName }"`, () => {
+				sinon.assert.calledOnce( spy );
```

Comment:
Most of the test logic resided in `beforeEach` hook. I'd move it inside `it`.

Reply:
Removing logic from `beforeEach` where we are firing the first input event may result with false-positive test for `input` event later (see a second `it`). Could you tell which parts you would move?

Reply:
Actually, it seems like there is no need to treat it any special as component value is not updated by `input` event (only spied). Looks like this test needs some refactoring thou. I will split it into 2 different test cases, as it's unreadable in the current format.